File: backend/app/api/v1/nasabah.py
from typing import Optional
from fastapi import APIRouter, Depends, Query, status
from app.core.deps import require_admin
from app.schemas.nasabah import (
    NasabahCreateRequest,
    NasabahStatusUpdateRequest,
    NasabahUpdateRequest,
)
from app.services.nasabah_service import nasabah_service
from app.services.transaction_service import transaction_service
from app.utils.responses import success_response

router = APIRouter(prefix="/admin/nasabah", tags=["Manajemen Nasabah (Admin)"])


@router.get("")
def list_nasabah(
    search: Optional[str] = Query(None, description="Pencarian nama, NIK, ID Nasabah, atau No. HP"),
    status: Optional[str] = Query(None, description="Filter status ACTIVE/INACTIVE"),
    page: int = Query(1, ge=1, description="Halaman"),
    page_size: int = Query(20, ge=1, le=1000, description="Jumlah item per halaman"),
    admin_user: dict = Depends(require_admin)
):
    """Daftar nasabah dengan pencarian, filter status, dan pagination."""
    items, total_items = nasabah_service.list_nasabah(
        search=search, status_filter=status, page=page, page_size=page_size
    )
    total_pages = (total_items + page_size - 1) // page_size if total_items > 0 else 0
    return success_response(
        data={
            "items": items,
            "pagination": {
                "page": page,
                "page_size": page_size,
                "total_items": total_items,
                "total_pages": total_pages,
            }
        }
    )


@router.post("", status_code=status.HTTP_201_CREATED)
def create_nasabah(data: NasabahCreateRequest, admin_user: dict = Depends(require_admin)):
    """Registrasi nasabah baru oleh Admin."""
    nasabah = nasabah_service.create_nasabah(
        data, registration_source="ADMIN", creator_id=admin_user["id"]
    )
    return success_response(
        data=nasabah,
        message="Nasabah berhasil didaftarkan.",
        status_code=status.HTTP_201_CREATED
    )


@router.get("/{nasabah_id}")
def get_nasabah_detail(nasabah_id: str, admin_user: dict = Depends(require_admin)):
    """Detail data nasabah beserta saldo."""
    nasabah = nasabah_service.get_nasabah_by_id(nasabah_id)
    return success_response(data=nasabah)


@router.put("/{nasabah_id}")
def update_nasabah(
    nasabah_id: str, data: NasabahUpdateRequest, admin_user: dict = Depends(require_admin)
):
    """Ubah data profil nasabah."""
    nasabah = nasabah_service.update_nasabah(nasabah_id, data, updater_id=admin_user["id"])
    return success_response(data=nasabah, message="Data nasabah berhasil diperbarui.")


@router.patch("/{nasabah_id}/status")
def update_nasabah_status(
    nasabah_id: str, data: NasabahStatusUpdateRequest, admin_user: dict = Depends(require_admin)
):
    """Aktifkan atau nonaktifkan status nasabah."""
    nasabah = nasabah_service.update_nasabah_status(nasabah_id, data.status, updater_id=admin_user["id"])
    return success_response(
        data=nasabah,
        message=f"Status nasabah berhasil diubah menjadi {data.status}."
    )


@router.get("/{nasabah_id}/balance")
def get_nasabah_balance(nasabah_id: str, admin_user: dict = Depends(require_admin)):
    """Mendapatkan saldo terakhir nasabah."""
    nasabah = nasabah_service.get_nasabah_by_id(nasabah_id)
    balance = nasabah_service.get_nasabah_balance(nasabah_id)
    return success_response(
        data={
            "nasabah_id": nasabah["id"],
            "customer_id": nasabah["customer_id"],
            "account_no": nasabah["account_no"],
            "name": nasabah["name"],
            "balance": balance,
        }
    )


@router.get("/{nasabah_id}/transactions")
def get_nasabah_transactions(
    nasabah_id: str,
    start_date: Optional[str] = Query(None, description="Tanggal awal (YYYY-MM-DD)"),
    end_date: Optional[str] = Query(None, description="Tanggal akhir (YYYY-MM-DD)"),
    type: Optional[str] = Query(None, description="SETOR atau TARIK"),
    category_id: Optional[str] = Query(None, description="ID Kategori"),
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
    admin_user: dict = Depends(require_admin)
):
    """Riwayat transaksi nasabah tertentu."""
    # Validate nasabah exists
    nasabah_service.get_nasabah_by_id(nasabah_id)
    items, total_items = transaction_service.list_transactions(
        nasabah_id=nasabah_id,
        start_date=start_date,
        end_date=end_date,
        type_filter=type,
        category_id=category_id,
        page=page,
        page_size=page_size
    )
    total_pages = (total_items + page_size - 1) // page_size if total_items > 0 else 0
    return success_response(
        data={
            "items": items,
            "pagination": {
                "page": page,
                "page_size": page_size,
                "total_items": total_items,
                "total_pages": total_pages,
            }
        }
    )

File: backend/src/routes/nasabah.ts
import { Hono } from 'hono';
import { AppEnv } from '../types/index.js';
import { nasabahService } from '../services/nasabahService.js';
import { authMiddleware } from '../middleware/auth.js';
import { requireAdmin } from '../middleware/roleGuard.js';
import { successResponse } from '../utils/response.js';
import { validateJson } from '../utils/validator.js';
import {
  nasabahCreateSchema,
  nasabahStatusUpdateSchema,
  nasabahUpdateSchema,
} from '../schemas/nasabah.schema.js';

export const nasabahRouter = new Hono<AppEnv>();

// All routes require ADMIN
nasabahRouter.use('*', authMiddleware, requireAdmin);

// GET /api/v1/admin/nasabah
nasabahRouter.get('/', async (c) => {
  const query = c.req.query();
  const result = await nasabahService.listNasabah({
    search: query.search,
    status: query.status,
    category: query.category,
    page: query.page ? parseInt(query.page, 10) : 1,
    page_size: query.page_size ? parseInt(query.page_size, 10) : 10,
  });
  return successResponse(c, result, 'Daftar nasabah berhasil dimuat.');
});

// POST /api/v1/admin/nasabah
nasabahRouter.post('/', validateJson(nasabahCreateSchema), async (c) => {
  const user = c.get('user');
  const data = c.req.valid('json' as any);
  const result = await nasabahService.createNasabah(data, 'ADMIN', user.id);
  return successResponse(c, result, 'Nasabah baru berhasil didaftarkan.', 201);
});

// GET /api/v1/admin/nasabah/:id
nasabahRouter.get('/:id', async (c) => {
  const id = c.req.param('id');
  const result = await nasabahService.getNasabahById(id);
  return successResponse(c, result, 'Detail nasabah berhasil dimuat.');
});

// PUT /api/v1/admin/nasabah/:id
nasabahRouter.put('/:id', validateJson(nasabahUpdateSchema), async (c) => {
  const user = c.get('user');
  const id = c.req.param('id');
  const data = c.req.valid('json' as any);
  const result = await nasabahService.updateNasabah(id, data, user.id);
  return successResponse(c, result, 'Data nasabah berhasil diperbarui.');
});

// PATCH /api/v1/admin/nasabah/:id/status
nasabahRouter.patch('/:id/status', validateJson(nasabahStatusUpdateSchema), async (c) => {
  const user = c.get('user');
  const id = c.req.param('id');
  const data = c.req.valid('json' as any);
  const result = await nasabahService.updateNasabahStatus(id, data.status, user.id);
  return successResponse(c, result, 'Status nasabah berhasil diubah.');
});

// GET /api/v1/admin/nasabah/:id/balance
nasabahRouter.get('/:id/balance', async (c) => {
  const id = c.req.param('id');
  const nasabah = await nasabahService.getNasabahById(id);
  const balance = await nasabahService.getNasabahBalance(nasabah.id);
  return successResponse(
    c,
    {
      nasabah_id: nasabah.id,
      customer_id: nasabah.customer_id,
      account_no: nasabah.account_no,
      name: nasabah.name,
      balance,
    },
    'Saldo nasabah berhasil dimuat.'
  );
});

// GET /api/v1/admin/nasabah/:id/transactions
nasabahRouter.get('/:id/transactions', async (c) => {
  const id = c.req.param('id');
  const query = c.req.query();
  const result = await nasabahService.getNasabahTransactions(id, {
    page: query.page ? parseInt(query.page, 10) : 1,
    page_size: query.page_size ? parseInt(query.page_size, 10) : 10,
    type: query.type,
  });
  return successResponse(c, result, 'Riwayat transaksi nasabah berhasil dimuat.');
});
